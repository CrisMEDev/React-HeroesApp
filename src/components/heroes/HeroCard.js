import React from 'react';
import { Link } from 'react-router-dom';

export const HeroCard = ({
    id,
    superhero,
    alter_ego,
    first_appearance,
    characters }) => {
    return (
        <div className="row g-4" style={{ maxWidth: 540 }}>
            <div className="col-md-6">
                <img
                src={`./assets/heroes/${id}.jpg`}
                className="img-fluid rounded-start"
                alt={superhero} />
            </div>
            <div className="col-md-6">
                <div className="card-body">
                    <h5 className="card-title">{superhero}</h5>
                    <p className="card-text">{alter_ego}</p>

                    {
                        ( alter_ego !== characters )
                        && 
                        <p className="card-text">{ characters }</p>
                    }

                    <p className="card-text"><small className="text-muted">{ first_appearance }</small></p>

                    <Link
                        className="btn btn-outline-dark"
                        to={ `/hero/${ id }` }
                    >
                        About...
                    </Link>

                </div>
            </div>
        </div>
    )
}
