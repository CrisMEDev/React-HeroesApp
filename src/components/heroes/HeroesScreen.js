import React from 'react';
import { useParams, Redirect } from 'react-router-dom';

import { getHeroeById } from '../../selectors/getHeroById';

export const HeroesScreen = () => {

    const { heroeId } = useParams();
    const hero = getHeroeById( heroeId );

    if ( !hero ) return <Redirect to="/" />;

    const {
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;

    return (
        <>
            <h1>Heroes Screen</h1>
        </>
    )
}
