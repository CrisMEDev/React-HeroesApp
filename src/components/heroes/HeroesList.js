import React, { useMemo } from 'react'

import { HeroCard } from './HeroCard';
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';

export const HeroesList = ({ publisher }) => {

    const heroes = useMemo(() => getHeroesByPublisher( publisher ), [ publisher ]);
    // const heroes = getHeroesByPublisher(publisher);

    return (
        <div className="row gy-8 animate__animated animate__fadeIn" >
            {
                heroes.map(hero => (
                    <HeroCard key={hero.id} {...hero} />
                ))
            }
        </div>
    )
}
