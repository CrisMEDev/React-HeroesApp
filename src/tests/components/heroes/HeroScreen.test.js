import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { HeroesScreen } from '../../../components/heroes/HeroesScreen';

describe('Pruebas en <HeroScreen />', () => {
    
    const historyMock = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    };
    
    
    test('Debe de mostrar el componente <Redirect /> si no hay argumentos en la URL', () => {
        const wrapper  = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroesScreen history={ historyMock } />
            </MemoryRouter>
        );
        
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('Redirect').exists() ).toBe(true);
        
    });

    test('Debe de mostrar un hero si el parÃ¡metro existe y se encuentra', () => {
        
        const wrapper  = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>

                {/* El argumento aqui debe ser como se especifico en DashboardRoutes */}
                <Route path="/hero/:heroeId" component={ HeroesScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find('.row').exists() ).toBe( true );
        
    });

    test('Debe de regresar a la pantalla principal con push', () => {
        
        const historyMock = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        };

        const wrapper  = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>

                {/* Se renderiza el componente en esta parte para mandar el historyMock */}
                <Route path="/hero/:heroeId" component={ (props) => <HeroesScreen history={historyMock} /> } />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();   // Ejecuta el handleReturn del boton

        expect( historyMock.push ).toHaveBeenCalledWith('/');
        expect( historyMock.goBack ).toHaveBeenCalledTimes( 0 );

    });
    
    test('Debe de regresar a la pantalla anterior con goBack', () => {
        
        const wrapper  = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>

                {/* Se renderiza el componente en esta parte para mandar el historyMock */}
                <Route path="/hero/:heroeId" component={ (props) => <HeroesScreen history={historyMock} /> } />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();   // Ejecuta el handleReturn del boton

        expect( historyMock.goBack ).toHaveBeenCalled();
        expect( historyMock.push ).toHaveBeenCalledTimes( 0 );
        
    });
    
    test('Debe de llamar el <Redirect /> si el hero no existe', () => {
        
        const wrapper  = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spidersafcdfbndsbfb']}>

                {/* Se renderiza el componente en esta parte para mandar el historyMock */}
                <Route path="/hero/:heroeId" component={ (props) => <HeroesScreen history={historyMock} /> } />
            </MemoryRouter>
        );

        expect( wrapper.text() ).toBe('');
        
        
    });
    
    
});
