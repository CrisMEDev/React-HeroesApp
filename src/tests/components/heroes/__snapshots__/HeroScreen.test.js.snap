// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen /> Debe de mostrar el componente <Redirect /> si no hay argumentos en la URL 1`] = `""`;
